drivers/sensor/pms7003/pms7003.c

```diff
@@ -0,0 +1,186 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* sensor pms7003.c - Driver for plantower PMS7003 sensor
+ * PMS7003 product: http://www.plantower.com/en/content/?110.html
+ * PMS7003 spec: http://aqicn.org/air/view/sensor/spec/pms7003.pdf
+ */
+
+#include <errno.h>
+
+#include <arch/cpu.h>
+#include <init.h>
+#include <kernel.h>
+#include <sensor.h>
+#include <stdlib.h>
+#include <string.h>
+#include <uart.h>
+
+#define SYS_LOG_DOMAIN "PMS7003"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_SENSOR_LEVEL
+#include <logging/sys_log.h>
+
+/* wait serial output with 1000ms timeout */
+#define CONFIG_PMS7003_SERIAL_TIMEOUT 1000
+
+struct pms7003_data {
+	struct device *uart_dev;
+	u16_t pm_1_0;
+	u16_t pm_2_5;
+	u16_t pm_10;
+};
+
+/**
+ * @brief wait for an array data from uart device with a timeout
+ *
+ * @param dev the uart device
+ * @param data the data array to be matched
+ * @param len the data array len
+ * @param timeout the timeout in milliseconds
+ * @return 0 if success; -ETIME if timeout
+ */
+int uart_wait_for(struct device *dev, uint8_t *data, int len, int timeout)
+{
+	int matched_size = 0;
+	s64_t timeout_time = k_uptime_get() + K_MSEC(timeout);
+
+	while (1) {
+		unsigned char c;
+
+		if (k_uptime_get() > timeout_time) {
+			return -ETIME;
+		}
+
+		if (uart_poll_in(dev, &c) == 0) {
+			if (c == data[matched_size]) {
+				matched_size++;
+
+				if (matched_size == len) {
+					break;
+				}
+			} else if (c == data[0]) {
+				matched_size = 1;
+			} else {
+				matched_size = 0;
+			}
+		}
+	}
+	return 0;
+}
+
+/**
+ * @brief read bytes from uart
+ *
+ * @param data the data buffer
+ * @param len the data len
+ * @param timeout the timeout in milliseconds
+ * @return 0 if success; -ETIME if timeout
+ */
+int uart_read_bytes(struct device *dev, uint8_t *data, int len, int timeout)
+{
+	int read_size = 0;
+	s64_t timeout_time = k_uptime_get() + K_MSEC(timeout);
+
+	while (1) {
+		unsigned char c;
+
+		if (k_uptime_get() > timeout_time) {
+			return -ETIME;
+		}
+
+		if (uart_poll_in(dev, &c) == 0) {
+			data[read_size++] = c;
+			if (read_size == len) {
+				break;
+			}
+		}
+	}
+	return 0;
+}
+
+static int pms7003_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct pms7003_data *drv_data = dev->driver_data;
+
+	/* sample output */
+	/* 42 4D 00 1C 00 01 00 01 00 01 00 01 00 01 00 01 01 92
+	 * 00 4E 00 03 00 00 00 00 00 00 71 00 02 06
+	 */
+
+	unsigned char pms7003_start_bytes[] = {0x42, 0x4d};
+	unsigned char pms7003_receive_buffer[30];
+
+	if (uart_wait_for(drv_data->uart_dev, pms7003_start_bytes,
+			  sizeof(pms7003_start_bytes),
+			  CONFIG_PMS7003_SERIAL_TIMEOUT) < 0) {
+		SYS_LOG_WRN("waiting for start bytes is timeout");
+		return -ETIME;
+	}
+
+	if (uart_read_bytes(drv_data->uart_dev, pms7003_receive_buffer, 30,
+			    CONFIG_PMS7003_SERIAL_TIMEOUT) < 0) {
+		return -ETIME;
+	}
+
+	drv_data->pm_1_0 =
+	    pms7003_receive_buffer[8] * 0xff + pms7003_receive_buffer[9];
```

Comment:
I'm guessing these are low and high bytes of the sample that you are combining here. If so, then you would need to do (high << 8) instead of (high * 0xFF).